sphinx/content/ch1.md

```diff
@@ -0,0 +1,1249 @@
+# Elegant NumPy: The Foundation of Scientific Python
+
+> [NumPy] is everywhere. It is all around us. Even now, in this very room.
+> You can see it when you look out your window or when you turn on your
+> television. You can feel it when you go to work... when you go to church...
+> when you pay your taxes.
+>
+> — Morpheus, *The Matrix*
+
+This chapter touches on some statistical functions in SciPy, but more than that, it focuses on exploring the NumPy array, a data structure that underlies almost all numerical scientific computation in Python.
+We will see how NumPy array operations enable concise and efficient code when manipulating numerical data.
+
+Our use case is using gene expression data from The Cancer Genome Atlas (TCGA) project to predict mortality in skin cancer patients.
+We will be working toward this goal throughout this chapter and the next one, learning about some key SciPy concepts along the way.
+Before we can predict mortality, we will need to normalize the expression data using a method called RPKM normalization.
+This allows the comparison of measurements between different samples and genes.
+(We will unpack what "gene expression" means in just a moment.)
+
+Let's start with a code snippet to tantalize you and introduce the ideas in this chapter.
+As we will do in each chapter, we open with a code sample that we believe epitomizes the elegance and power of a particular function from the SciPy ecosystem.
+In this case, we want to highlight NumPy's vectorization and broadcasting rules, which allow us to manipulate and reason about data arrays very efficiently.
+
+
+```{code-block} python
+---
+caption: 'By the end of the chapter, this will all make sense!'
+name: 'code:intro'
+---
+def rpkm(counts, lengths):
+    """Calculate reads per kilobase transcript per million reads.
+
+    RPKM = (10^9 * C) / (N * L)
+
+    Where:
+    C = Number of reads mapped to a gene
+    N = Total mapped reads in the experiment
+    L = Exon length in base pairs for a gene
+
+    Parameters
+    ----------
+    counts: array, shape (N_genes, N_samples)
+        RNAseq (or similar) count data where columns are individual samples
+        and rows are genes.
+    lengths: array, shape (N_genes,)
+        Gene lengths in base pairs in the same order
+        as the rows in counts.
+
+    Returns
+    -------
+    normed : array, shape (N_genes, N_samples)
+        The RPKM normalized counts matrix.
+    """
+    # First, convert counts to float to avoid overflow when multiplying by
+    # 1e9 in the RPKM formula
+    C = counts.astype(float)
+    N = np.sum(C, axis=0)  # sum each column to get total reads per sample
+    L = lengths
+
+    normed = 1e9 * C / (N[np.newaxis, :] * L[:, np.newaxis])
+
+    return(normed)
+```
+
+The `rpkm` function defined in {numref}`code:intro` illustrates some of the
```

Comment:
Why did they choose not to use the ReST syntax for this? I.e. :numref:`...` ?

Reply:
MyST provides a markdown syntax for *both* rST roles and directives. I can't speak to the specific design choices, but I do think the {}'s are more consistent with the rest of the MyST syntax. Also, I think it could be confusing to have the directives have their own, new syntax, whereas the roles still use rST syntax.